theano/gradient.py

```diff
@@ -1201,58 +1201,62 @@ def try_to_copy_if_needed(var):
-                            msg += "of zeros. It is not NullType or "
-                            msg += "DisconnectedType and theano can't "
-                            msg += "simplify it to a constant, so it's not "
-                            msg += "verifiably zeros."
-
-                            msg = msg % (str(node.op), str(term),
-                                         str(type(term)), i)
-
-                        if is_zero == 'no':
-                            msg = "%s.grad returned %s of type %s for input"
-                            msg += " %d. Since this input is only connected "
-                            msg += "to integer-valued outputs, it should "
-                            msg += "evaluate to zeros, but it evaluates to"
-                            msg += "%s."
-
-                            msg % (node.op, term, type(term), i,
+                            msg = "%s.grad returned %s of type %s for input" \
+                                " %d. This input's only connections to " \
+                                "the cost through this op are via " \
+                                "integer-valued outputs so it should be " \
+                                "NullType, DisconnectedType, or some form " \
+                                "of zeros. It is not NullType or " \
+                                "DisconnectedType and theano can't " \
+                                "simplify it to a constant, so it's not " \
+                                "verifiably zeros."
+
+                            msg %= (node.op, term, type(term), i)
+
+                        elif is_zero == 'no':
+                            msg = "%s.grad returned %s of type %s for input" \
+                                " %d. Since this input is only connected " \
+                                "to integer-valued outputs, it should " \
+                                "evaluate to zeros, but it evaluates to" \
+                                "%s."
+
+                            msg %= (node.op, term, type(term), i,
                                    theano.get_scalar_constant_value(term))
 
                             raise ValueError(msg)
 
             # Check that op.connection_pattern matches the connectivity
             # logic driving the op.grad method
-            for i, packed in enumerate(zip(inputs, input_grads,
-                                           inputs_connected)):
-                ipt, ig, connected = packed
+            for i, (ipt, ig, connected) in enumerate(
+                zip(inputs, input_grads, inputs_connected)):
                 actually_connected = \
                     not isinstance(ig.type, DisconnectedType)
 
+                if isinstance(node.op, theano.OpFromGraph):
+                    ov = node.op._grad_op_overrides_l[i]
+                    if ov is not None:
+                        connected &= not isinstance(
+                            ov.type, DisconnectedType)
```

Comment:
I think the `connection_pattern` should really be taken as "should gradients flow through there?", in particular "if there's an invalid output gradient, should the input gradient be invalid too?" and "if this is the only path between the cost and the param, is the user likely to have confused some variables?".
If the user explicitly overrode the gradient to be an actual expression, I say the answer to those questions would be yes, and the connection pattern should be `True`.
If the user explicitly overrode the gradient to be `DisconnectedType()`, I say the answer is no, and the connection pattern should be `False`.
If it is not overriden, then use `io_connection_pattern`.

That way, I think the current check still makes sense.